File: toggly-go/toggly/eval/builtin.go
package eval

import (
	"strings"
	"time"
)

// AlwaysOnEvaluator implements the AlwaysOn filter.
type AlwaysOnEvaluator struct{}

func (AlwaysOnEvaluator) Evaluate(featureKey string, params map[string]any, ctx Context) (bool, error) {
	_, _, _ = featureKey, params, ctx
	return true, nil
}

// AlwaysOffEvaluator implements the AlwaysOff filter (e.g. synthesized from evaluated variants).
type AlwaysOffEvaluator struct{}

func (AlwaysOffEvaluator) Evaluate(featureKey string, params map[string]any, ctx Context) (bool, error) {
	_, _, _ = featureKey, params, ctx
	return false, nil
}

// PercentageEvaluator implements Percentage.
type PercentageEvaluator struct{}

func (PercentageEvaluator) Evaluate(featureKey string, params map[string]any, ctx Context) (bool, error) {
	pct, ok := asFloat(params, "Value")
	if !ok {
		pct, _ = asFloat(params, "Percentage")
	}
	if pct <= 0 {
		return false, nil
	}
	if pct >= 100 {
		return true, nil
	}

	if ctx.Identity == "" {
		return false, nil
	}
	return ComputePercentile(ctx.Identity, featureKey) < pct, nil
}

// TimeWindowEvaluator implements TimeWindow.
type TimeWindowEvaluator struct {
	Now func() time.Time
}

func (t TimeWindowEvaluator) Evaluate(featureKey string, params map[string]any, ctx Context) (bool, error) {
	// Definitions parity: each side is optional; missing side is unconstrained;
	// neither Start nor End → true; invalid present side fails closed.
	_, _ = featureKey, ctx
	startS, _ := asString(params, "Start")
	endS, _ := asString(params, "End")
	now := time.Now().UTC()
	if t.Now != nil {
		now = t.Now().UTC()
	}

	if startS != "" {
		start, ok := parseTime(startS)
		if !ok || now.Before(start) {
			return false, nil
		}
	}

	if endS != "" {
		end, ok := parseTime(endS)
		if !ok || now.After(end) {
			return false, nil
		}
	}

	return true, nil
}

// TargetingEvaluator implements Targeting.
type TargetingEvaluator struct{}

func (TargetingEvaluator) Evaluate(featureKey string, params map[string]any, ctx Context) (bool, error) {
	ignoreCase, ok := asBool(params, "IgnoreCase")
	if !ok {
		ignoreCase = true // Definitions default
	}

	identity := ctx.Identity
	groups := ctx.Groups

	exclusionUsers := collectIndexedValues(params, []string{"Audience.Exclusion.Users", "Audience:Exclusion:Users"})
	if identity != "" && contains(exclusionUsers, identity, ignoreCase) {
		return false, nil
	}
	exclusionGroups := collectIndexedValues(params, []string{"Audience.Exclusion.Groups", "Audience:Exclusion:Groups"})
	if len(groups) > 0 {
		for _, eg := range exclusionGroups {
			if contains(groups, eg, ignoreCase) {
				return false, nil
			}
		}
	}

	if identity != "" {
		users := collectIndexedValues(params, []string{"Audience.Users", "Audience:Users"})
		if contains(users, identity, ignoreCase) {
			return true, nil
		}
	}

	if len(groups) > 0 {
		audienceGroups := collectIndexedValues(params, []string{"Audience.Groups", "Audience:Groups"})
		for _, g := range groups {
			if contains(audienceGroups, g, ignoreCase) {
				return true, nil
			}
		}
	}

	pct, ok := asFloat(params, "Audience.DefaultRolloutPercentage")
	if !ok {
		pct, _ = asFloat(params, "Percentage")
	}
	if pct <= 0 {
		return false, nil
	}
	if pct >= 100 {
		return true, nil
	}
	if identity == "" {
		return false, nil
	}

	return ComputePercentile(identity, featureKey) < pct, nil
}

func parseTime(s string) (time.Time, bool) {
	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
		return t, true
	}
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t, true
	}
	return time.Time{}, false
}

func contains(list []string, val string, ignoreCase bool) bool {
	for _, s := range list {
		if ignoreCase {
			if strings.EqualFold(s, val) {
				return true
			}
			continue
		}
		if s == val {
			return true
		}
	}
	return false
}

func asString(params map[string]any, key string) (string, bool) {
	v, ok := params[key]
	if !ok {
		return "", false
	}
	s, ok := asStringValue(v)
	return s, ok
}

func asStringValue(v any) (string, bool) {
	s, ok := v.(string)
	if ok {
		return s, true
	}
	return "", false
}

// NOTE: If you need a non-deterministic rollout, register a custom filter and
// implement your own sampling strategy.

File: toggly-eval/src/builtin.ts
import type { FilterEvaluator } from './types'
import { computePercentile } from './hash'
import { registerSegmentFilters } from './segment'
import { asBool, asFloat, asString, collectIndexedValues } from './params'

export { asBool, asFloat, asString, collectIndexedValues } from './params'

function parseTime(s: string): Date | undefined {
  const t = Date.parse(s)
  if (Number.isNaN(t)) {
    return undefined
  }
  return new Date(t)
}

function contains(list: string[], val: string, ignoreCase: boolean): boolean {
  for (const s of list) {
    if (ignoreCase) {
      if (s.toLowerCase() === val.toLowerCase()) {
        return true
      }
    } else if (s === val) {
      return true
    }
  }
  return false
}

export const alwaysOn: FilterEvaluator = () => true
export const alwaysOff: FilterEvaluator = () => false

export const percentage: FilterEvaluator = (featureKey, params, ctx) => {
  let pct = asFloat(params, 'Value')
  if (pct === undefined) {
    pct = asFloat(params, 'Percentage')
  }
  if (pct === undefined || pct <= 0) {
    return false
  }
  if (pct >= 100) {
    return true
  }
  if (!ctx.identity) {
    return false
  }
  return computePercentile(ctx.identity, featureKey) < pct
}

let timeWindowNow: (() => Date) | undefined

/** Test hook to pin TimeWindow "now". */
export function setTimeWindowNow(fn: (() => Date) | undefined): void {
  timeWindowNow = fn
}

export const timeWindow: FilterEvaluator = (_featureKey, params, _ctx) => {
  // Definitions parity: each side is optional; missing side is unconstrained;
  // neither Start nor End → true; invalid present side fails closed.
  const startS = asString(params, 'Start')
  const endS = asString(params, 'End')
  const now = (timeWindowNow?.() ?? new Date()).getTime()

  if (startS) {
    const start = parseTime(startS)
    if (!start || now < start.getTime()) {
      return false
    }
  }

  if (endS) {
    const end = parseTime(endS)
    if (!end || now > end.getTime()) {
      return false
    }
  }

  return true
}

export const targeting: FilterEvaluator = (featureKey, params, ctx) => {
  // Match Definitions default: IgnoreCase defaults to true when unset.
  const ignoreCase = asBool(params, 'IgnoreCase') ?? true
  const identity = ctx.identity ?? ''
  const groups = ctx.groups ?? []

  const exclusionUsers = collectIndexedValues(params, [
    'Audience.Exclusion.Users',
    'Audience:Exclusion:Users',
  ])
  if (identity && contains(exclusionUsers, identity, ignoreCase)) {
    return false
  }

  const exclusionGroups = collectIndexedValues(params, [
    'Audience.Exclusion.Groups',
    'Audience:Exclusion:Groups',
  ])
  if (
    groups.length > 0 &&
    exclusionGroups.some((g) => contains(groups, g, ignoreCase))
  ) {
    return false
  }

  if (identity) {
    const users = collectIndexedValues(params, [
      'Audience.Users',
      'Audience:Users',
    ])
    if (contains(users, identity, ignoreCase)) {
      return true
    }
  }

  if (groups.length > 0) {
    const audienceGroups = collectIndexedValues(params, [
      'Audience.Groups',
      'Audience:Groups',
    ])
    for (const g of groups) {
      if (contains(audienceGroups, g, ignoreCase)) {
        return true
      }
    }
  }

  let pct = asFloat(params, 'Audience.DefaultRolloutPercentage')
  if (pct === undefined) {
    pct = asFloat(params, 'Percentage')
  }
  if (pct === undefined || pct <= 0) {
    return false
  }
  if (pct >= 100) {
    return true
  }
  if (!identity) {
    return false
  }
  return computePercentile(identity, featureKey) < pct
}

export function createDefaultRegistry(): Map<string, FilterEvaluator> {
  const reg = new Map<string, FilterEvaluator>()
  const register = (names: string[], ev: FilterEvaluator) => {
    for (const name of names) {
      reg.set(name, ev)
    }
  }
  register(['AlwaysOn'], alwaysOn)
  register(['AlwaysOff'], alwaysOff)
  register(['Percentage', 'Microsoft.Percentage'], percentage)
  register(['TimeWindow', 'Microsoft.TimeWindow'], timeWindow)
  register(['Targeting', 'Microsoft.Targeting'], targeting)
  registerSegmentFilters(reg)
  return reg
}

/** Exported for tests that need bucket values. */
export { computePercentile, identityBucket, rolloutBucket } from './hash'
